mlflow/data/pandas_dataset.py

```diff
@@ -0,0 +1,116 @@
+from mlflow.types import Schema
+from mlflow.types.utils import _infer_schema
+
+
+class PandasDataset(Dataset):
+    def __init__(
+        self,
+        df: pd.DataFrame,
+        source: FileSystemDatasetSource,
+        name: Optional[str] = None,
+        digest: Optional[str] = None,
+    ):
+        """
+        TODO: Pandas docs
+        """
+        self._df = df
+        super().__init__(source=source, name=name, digest=digest)
+
+    def _compute_digest(self) -> str:
+        """
+        Computes a digest for the dataset. Called if the user doesn't supply
+        a digest when constructing the dataset.
+        """
+        MAX_ROWS = 10000
+
+        # drop object columns
+        df = self._df.select_dtypes(exclude=["object"])
+        trimmed_df = df.head(MAX_ROWS)
+        # hash trimmed dataframe contents
+        md5 = hashlib.md5(pd.util.hash_pandas_object(trimmed_df).values)
+        # hash dataframe dimensions
+        n_rows = len(df)
+        md5.update(np.int64(n_rows))
+        # hash column names
+        columns = df.columns
+        for x in columns:
+            md5.update(x.encode())
+        # TODO: Make this a normalize_hash function (truncation)
+        return md5.hexdigest()[:8]
+
+    def _to_dict(self, base_dict: Dict[str, str]) -> Dict[str, str]:
+        """
+        :param base_dict: A string dictionary of base information about the
+                          dataset, including: name, digest, source, and source
+                          type.
+        :return: A string dictionary containing the following fields: name,
+                 digest, source, source type, schema (optional), profile
+                 (optional).
+        """
+        base_dict.update(
+            {
+                "schema": json.dumps({"mlflow_colspec": self.schema.to_dict()}),
+                "profile": json.dumps(self.profile),
+            }
+        )
+        return base_dict
+
+    @property
+    def source(self) -> FileSystemDatasetSource:
+        """
+        TODO: Pandas docs
+        """
+        return self._source
+
+    @property
+    def df(self) -> pd.DataFrame:
+        return self._df
+
+    @property
+    def profile(self) -> Optional[Any]:
+        """
+        TODO: Pandas docs
+        """
+        return {
+            "num_rows": len(self._df),
+            "num_elements": int(self._df.size),
+        }
+
+    @property
+    def schema(self) -> Schema:
+        """
+        An MLflow ColSpec schema representing the columnar dataset
+        """
+        return _infer_schema(self._df)
+
+
+def from_pandas(
+    df: pd.DataFrame, source: str, name: Optional[str] = None, digest: Optional[str] = None
+) -> PandasDataset:
+    """
+    :param df: A Pandas DataFrame
+    :param source: The source from which the DataFrame was derived. E.g. a Spark table
+                    name, a delta table name with version, a distributed filesystem path or
+                    URI, etc.
```

Comment:
For NumPy we say filesystem, but here we say pretty much anything. I think we should support the following:

- Filesystem dataset sources, including cloud blob storage (covered by `FileSystemDatasetSource`)
- Delta table sources (Pandas DF might come from a Delta table)
- Spark table sources (Pandas DF might come from a Spark table)

I also think we should support HTTP/S; I'll file a follow-up ticket to implement that dataset source :)